### ISSUE TITLE
Duplicate Queries / UNSIGNED Attribute Issue

### ISSUE THREAD

**Issue by xlxAciDxlx** [typeorm/typeorm#2861]
**Issue type:**

[x] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[x] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[x] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

- Setup TypeORM
- Create Entity without UNSIGNED Primary Column:
```ts
@Entity({ name: 'commands' })
export class Command extends BaseEntity {
	@PrimaryGeneratedColumn()
	commandId: number;

	@Column()
	description: string;

	@Column()
	name: string;

	@Column()
	response: string;

	@OneToMany(type => Service, service => service.serviceId)
	@JoinTable()
	service: Service | null;

	@Column()
	serviceKey: string;
}
```
- Modify `@PrimaryGeneratedColumn()` to be UNSIGNED:
```ts
	@PrimaryGeneratedColumn({ unsigned: true })
```

That works fine. The issue I have is now every time I start my server, it runs the following command _every time_ and twice:
```sql
ALTER TABLE `commands` CHANGE `commandId` `commandId` int UNSIGNED NOT NULL AUTO_INCREMENT
```

Shouldn't that only run _once_ after the change is made? I also have no clue what's causing it to be run twice, but it does it with _every_ table I set to UNSIGNED.

**Comment by xlxAciDxlx** [comment]
Yes, I do have `synchronize` enabled. Also, I should've noted I'm running this in development as I'm currently working on planning out my schemas and such. So should I still only use `synchronize` once? Also, it seems odd that those queries are being run twice, I can also confirm it with `CREATE TABLE` queries. Which are then followed by errors that, as expected, the table already exists.

**Comment by elthrasher** [comment]
I don't have too much experience with `synchronize`. Maybe you also want to set `dropSchema`?

**Comment by xlxAciDxlx** [comment]
That helps avoid the error, but that doesn't help the underlying issue - being the double queries (does it for schema selection, dropping _some_ tables, and creating _some_ tables).

**Comment by elthrasher** [comment]
I tried running a few of the unit tests in typeorm, but none of them quite fit this case. Want to create a PR that proves the case? Would make it easier to debug.

**Comment by xlxAciDxlx** [comment]
Alright, so maybe I'm at fault here. I created a barebones repro case with minimal tables and configuration changes from default TypeORM setup (as to ensure I wasn't maybe configuring something that messed it up) - and it doesn't reproduce... The issue is, the "broken" case I'm having can't be publicized as it's not yet a public project.

Is there anything else I can test on the "broken" use case to debug this further?

**Comment by elthrasher** [comment]
Did you include the join table and other relationships in your sample? If there's a bug, I'll bet it's got something to do with how typeorm traverses your schema.

**Comment by xlxAciDxlx** [comment]
I went as far as to copy my _exact_ entity classes to the test repro directory, and everything ran fine - no errors, no double queries. It did still execute the query every time, but if that goes away when `synchronize` is false then that's fine (though still doesn't seem desired after the first synchronization).

**Comment by elthrasher** [comment]
Yes, when you remove synchronize, you'll stop seeing those things. It's really only useful for development. If you are deploying your app to production and want it to create the schema automatically, migrations are the right tool for that. Or of course you can manage your database outside typeorm.

**Comment by xlxAciDxlx** [comment]
Yeah, that was the whole reason I still had `synchronize` enabled - I'm still working on hashing everything out. The only problem is using `dropSchema` isn't a viable "solution" as that prevents me from being able to reliably actually test this as an API (at least while I'm still figuring out how everything should be laid out).

**Comment by xlxAciDxlx** [comment]
That would work for production (if need be, but that'd defeat the purpose of migrations), but in development that wouldn't at all be practical, unfortunately.

**Comment by xlxAciDxlx** [comment]
So, a _really_ weird development. The double queries seem to be caused by setting a custom logger. I'd create a reproduction repository, but everything you need to try it in your own project is here.

### ormconfig.json
Add `logging: "all"`

### Setting Logger (before `createExpressServer` from `routing-controllers`)
```ts
getConnectionOptions().then(connectionOptions => {
	return createConnection(
		Object.assign(connectionOptions, {
			logger: new CustomLogger()
		})
	);
});
```

### Logger Class
```ts
import chalk from 'chalk';
import { EOL } from 'os';
import { Logger, QueryRunner } from 'typeorm';

export class CustomLogger implements Logger {
	private colorizeQuery(query: string): string {
		// ...
	}

	log(level: 'error' | 'log' | 'info' | 'warn', message: any, queryRunner?: QueryRunner) {}

	logMigration(message: string, queryRunner?: QueryRunner) {
		console.info(`[Database] [MIGRATION] ${message}`);
	}

	logQuery(query: string, parameters?: any[], queryRunner?: QueryRunner) {
		console.info('[Database] [QUERY] ' + this.colorizeQuery(query));
	}

	logQueryError(error: string, query: string, parameters?: any[], queryRunner?: QueryRunner) {
		console.error(chalk.redBright(`[Database] [ERROR] ${error}`));
	}

	logQuerySlow(time: number, query: string, parameters?: any[], queryRunner?: QueryRunner) {
		let slowTime = `${time}MS`;
		if (time > 1000) {
			slowTime = time / 1000 + ' seconds';
		}

		console.warn(chalk.yellowBright(`[Database] [WARN] Slow query detected! Took ${slowTime} to execute!` + EOL + `Query: ${query}`));
	}

	logSchemaBuild(message: string, queryRunner?: QueryRunner) {}
}

```

**Comment by nicolae536** [comment]
I have the same issue replacing with a custom logger is logging twice to the console.
